drivers/sensor/stm32_temp/stm32_temp.c

```diff
@@ -88,11 +88,37 @@ static int stm32_temp_channel_get(const struct device *dev, enum sensor_channel
 	}
 
 #if HAS_CALIBRATION
+
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	/*
+	 * TS_CAL are 16bit values (cf Data Sheet), they are divided to match the 12bit
+	 * resolution of the data_raw.
+	 * The 12bit resolution is set by the asp.resolution in stm32_temp_init()
+	 */
+	temp = ((float)data->raw * cfg->tsv_mv) / cfg->cal_vrefanalog;
+	temp -= (*cfg->cal1_addr >> 4);
+	temp *= cfg->cal2_temp - cfg->cal1_temp;
+	temp /= (*cfg->cal2_addr >> 4) - (*cfg->cal1_addr >> 4);
+	temp += cfg->cal_offset;
+
+#elif defined(CONFIG_SOC_SERIES_STM32U5X)
+	/*
+	 * TS_CAL are 14bit values (cf Data Sheet), they are divided to match the 12bit
+	 * resolution of the data_raw.
+	 * The 12bit resolution is set by the asp.resolution in stm32_temp_init()
+	 */
+	temp = ((float)data->raw * cfg->tsv_mv) / cfg->cal_vrefanalog;
+	temp -= (*cfg->cal1_addr >> 2);
+	temp *= cfg->cal2_temp - cfg->cal1_temp;
+	temp /= (*cfg->cal2_addr >> 2) - (*cfg->cal1_addr >> 2);
+	temp += cfg->cal_offset;
+#else
 	temp = ((float)data->raw * cfg->tsv_mv) / cfg->cal_vrefanalog;
 	temp -= *cfg->cal1_addr;
 	temp *= (cfg->cal2_temp - cfg->cal1_temp);
 	temp /= (*cfg->cal2_addr - *cfg->cal1_addr);
 	temp += cfg->cal_offset;
+#endif
```

Comment:
Can we abstract this difference using device tree ?

Reply:
Using the soc DTS  should introduce a new property <ADC resolution for TS_CAL> for each ADC with <st_stm32_temp_cal>.
The data TS_CAL1 (*cfg->cal1_addr) and TS_CAL2 (*cfg->cal2_addr) are stored in the system memory of the device. Depending on the SOC, the ADC is 12 or 14 or 16 bits wide and the TS_CAL1 and TS_CAL2 are measured at this max resolution. 
On the other hand the stm32_temp.c is always converting with 12bit resolution :
```
*asp = (struct adc_sequence){
		.resolution = 12U
}
```
That's why the shift is required.

Reply:
Could also be done with a "custom flag" (like in  the  include/zephyr/dt-bindings/pwm/stm32_pwm.h with PWM_POLARITY_NORMAL  / PWM_STM32_COMPLEMENTARY)
--> TS_CAL_STM32_16BIT / TS_CAL_STM32_14BIT / TS_CAL_STM32_12BIT

Reply:
> Could also be done with a "custom flag" (like in the include/zephyr/dt-bindings/pwm/stm32_pwm.h with PWM_POLARITY_NORMAL / PWM_STM32_COMPLEMENTARY) --> TS_CAL_STM32_16BIT / TS_CAL_STM32_14BIT as follows:
> 
> ```
> 		die_temp: dietemp {
> 			compatible = "st,stm32-temp-cal";
> 			ts-cal1-addr = <0x0BFA0710 | TS_CAL_STM32_16BIT>;
> 			ts-cal2-addr = <0x0BFA0742 | TS_CAL_STM32_16BIT>;
> ```

Using a flag on a 32 bits addr field looks a bit fragile. Can't we add a new `ts-cal-resolution` field ?